lib/src/string_compare.dart

```diff
@@ -2,13 +2,20 @@
 /// source code.
 ///
 /// This mostly follows the same rules as `String.trim()` because that's what
-/// dart_style uses to trim trailing whitespace as well as considering `,` to
-/// be a whitespace character since the formatter will add and remove trailing
-/// commas.
+/// dart_style uses to trim trailing whitespace.
+///
+/// This function treats `,` as a whitespace character since the formatter will
+/// add and remove trailing commas. It treats, `[`, `]`, `{`, and `}` as
+/// whitespace characters because the formatter may move a comment if it
+/// appears near the closing delimiter of an optional parameter section.
```

Comment:
Consider having a separate stack of open `[{`s and check that they all occur in both strings, in the correct nesting, even if we don't know that they occur in the correct places.

Or considering recognizing comments instead of `[]{}`, and ignoring the position of them, while checking that they still all occur, in the same order. Seems less error-prone than ignoring order of significan syntax.

Reply:
I agree this function is getting weirder. Added a TODO comment to look into it more. There is some tension in that:

* This function exists entirely as a sanity check for the formatter itself. It is only useful when there are bugs in the formatter: it can sometimes prevent user code from getting discarded. This *has* occurred several times in the wild, so it's a useful sanity check. But 99.999% of the time, it doesn't add any user value.

* It is fairly heavy in terms of performance. It walks every character of the original and formatted output, so it adds a noticeable amount of overhead to every run of the formatter.

Finding the right balance between a sanity check that does catch bugs and one that doesn't add too much overhead is hard.